File: Examples/C/Examples/FileConverter.c
#include "../../../x-IMU3-API/C/Ximu3.h"
#include "../Helpers.h"
#include <inttypes.h>
#include <stdio.h>

static void Callback(const XIMU3_FileConverterProgress progress, void* context);

static void PrintProgress(const XIMU3_FileConverterProgress progress);

void FileConverter()
{
    const char* destination = "C:/";
    const char* name = "File Conversion Example";
    const char* file_paths[] = { "C:/file.ximu3" }; // replace with actual file path
    const int numberOfFiles = sizeof(file_paths) / sizeof(file_paths[0]);

    if (YesOrNo("Use async implementation?"))
    {
        XIMU3_FileConverter* const file_converter = XIMU3_file_converter_new(destination, name, file_paths, numberOfFiles, Callback, NULL);

        Wait(60);

        XIMU3_file_converter_free(file_converter);
    }
    else
    {
        PrintProgress(XIMU3_file_converter_convert(destination, name, file_paths, numberOfFiles));
    }
}

static void Callback(const XIMU3_FileConverterProgress progress, void* context)
{
    PrintProgress(progress);
}

static void PrintProgress(const XIMU3_FileConverterProgress progress)
{
    printf("%s, %0.1f%%, %" PRIu64 " of %" PRIu64 " bytes\n",
           XIMU3_file_converter_status_to_string(progress.status),
           progress.percentage,
           progress.bytes_processed,
           progress.bytes_total);
    // printf("%s\n", XIMU3_file_converter_progress_to_string(progress)); // alternative to above
}

File: Examples/Rust/src/examples/file_converter.rs
use ximu3::file_converter::*;

pub fn run() {
    // Blocking
    let destination = "C:/Users/Public/";
    let name_blocking = "x-IMU3 File Conversion Example Blocking";
    let file_paths = vec!["C:/Users/Public/x-IMU3 Example File.ximu3"]; // replace with actual file path

    let result = FileConverter::convert(destination, name_blocking, file_paths.clone());

    if let Err(error) = result {
        println!("File converter failed: {error}");
    }

    // Non-blocking
    let name_non_blocking = "x-IMU3 File Conversion Example Non-Blocking";

    let closure = Box::new(|progress: FileConverterProgress| {
        println!("{}, {:.1}%, {} of {} bytes", progress.status, progress.percentage, progress.bytes_processed, progress.bytes_total);
        // println!("{progress}"); // alternative to above
    });

    let file_converter = FileConverter::new(destination, name_non_blocking, file_paths, closure);

    if let Err(error) = file_converter {
        println!("File converter failed: {error}");
    }

    std::thread::sleep(std::time::Duration::from_secs(60));
}
